File: ecmascript/ecma_string_table.cpp
/*
 * Copyright (c) 2021 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "ecmascript/ecma_string_table.h"

#include "ecmascript/ecma_string-inl.h"
#include "ecmascript/ecma_vm.h"
#include "ecmascript/js_thread.h"
#include "ecmascript/mem/c_string.h"
#include "ecmascript/object_factory.h"

namespace panda::ecmascript {
EcmaStringTable::EcmaStringTable(const EcmaVM *vm) : vm_(vm) {}

EcmaString *EcmaStringTable::GetString(const JSHandle<EcmaString> &firstString, const JSHandle<EcmaString> &secondString) const
{
    uint32_t hashCode = firstString->GetHashcode();
    hashCode = secondString->ComputeHashcode(hashCode);
    auto range = table_.equal_range(hashCode);
    auto item = range.first;
    for (; item != range.second; ++item) {
        auto foundString = item->second;
        if (foundString->EqualToSplicedString(*firstString, *secondString)) {
            return foundString;
        }
    }
    return nullptr;
}

EcmaString *EcmaStringTable::GetString(const uint8_t *utf8Data, uint32_t utf8Len, bool canBeCompress) const
{
    uint32_t hashCode = EcmaString::ComputeHashcodeUtf8(utf8Data, utf8Len, canBeCompress);
    auto range = table_.equal_range(hashCode);
    auto item = range.first;
    for (; item != range.second; ++item) {
        auto foundString = item->second;
        if (EcmaString::StringsAreEqualUtf8(foundString, utf8Data, utf8Len, canBeCompress)) {
            return foundString;
        }
    }
    return nullptr;
}

EcmaString *EcmaStringTable::GetString(const uint16_t *utf16Data, uint32_t utf16Len) const
{
    uint32_t hashCode = EcmaString::ComputeHashcodeUtf16(const_cast<uint16_t *>(utf16Data), utf16Len);
    auto range = table_.equal_range(hashCode);
    auto item = range.first;
    for (; item != range.second; ++item) {
        auto foundString = item->second;
        if (EcmaString::StringsAreEqualUtf16(foundString, utf16Data, utf16Len)) {
            return foundString;
        }
    }
    return nullptr;
}

EcmaString *EcmaStringTable::GetString(EcmaString *string) const
{
    auto range = table_.equal_range(string->GetHashcode());
    auto item = range.first;
    for (; item != range.second; ++item) {
        auto foundString = item->second;
        if (EcmaString::StringsAreEqual(foundString, string)) {
            return foundString;
        }
    }
    return nullptr;
}

void EcmaStringTable::InternString(EcmaString *string)
{
    if (string->IsInternString()) {
        return;
    }
    table_.insert(std::pair<uint32_t, EcmaString *>(string->GetHashcode(), string));
    string->SetIsInternString();
}

void EcmaStringTable::InternEmptyString(EcmaString *emptyStr)
{
    InternString(emptyStr);
}

EcmaString *EcmaStringTable::GetOrInternString(const JSHandle<EcmaString> &firstString, const JSHandle<EcmaString> &secondString)
{
    if (firstString->IsInternString() && secondString->IsInternString()) {
        EcmaString *concatString = GetString(firstString, secondString);
        if (concatString != nullptr) {
            return concatString;
        }
    }
    EcmaString *concatString = EcmaString::Concat(firstString, secondString, vm_);

    InternString(concatString);
    return concatString;
}

EcmaString *EcmaStringTable::GetOrInternString(const uint8_t *utf8Data, uint32_t utf8Len, bool canBeCompress)
{
    EcmaString *result = GetString(utf8Data, utf8Len, canBeCompress);
    if (result != nullptr) {
        return result;
    }

    result = EcmaString::CreateFromUtf8(utf8Data, utf8Len, vm_, canBeCompress);
    InternString(result);
    return result;
}

EcmaString *EcmaStringTable::GetOrInternString(const uint16_t *utf16Data, uint32_t utf16Len, bool canBeCompress)
{
    EcmaString *result = GetString(utf16Data, utf16Len);
    if (result != nullptr) {
        return result;
    }

    result = EcmaString::CreateFromUtf16(utf16Data, utf16Len, vm_, canBeCompress);
    InternString(result);
    return result;
}

EcmaString *EcmaStringTable::GetOrInternString(EcmaString *string)
{
    if (string->IsInternString()) {
        return string;
    }

    EcmaString *result = GetString(string);
    if (result != nullptr) {
        return result;
    }
    InternString(string);
    return string;
}

void EcmaStringTable::SweepWeakReference(const WeakRootVisitor &visitor)
{
    for (auto it = table_.begin(); it != table_.end();) {
        auto *object = it->second;
        auto fwd = visitor(object);
        if (fwd == nullptr) {
            LOG(DEBUG, GC) << "StringTable: delete string " << std::hex << object
                           << ", val = " << ConvertToString(object);
            table_.erase(it++);
        } else if (fwd != object) {
            it->second = static_cast<EcmaString *>(fwd);
            ++it;
            LOG(DEBUG, GC) << "StringTable: forward " << std::hex << object << " -> " << fwd;
        } else {
            ++it;
        }
    }
}
}  // namespace panda::ecmascript

File: test/moduletest/ecmastringtable/ecmastringtable.js
/*
 * Copyright (c) 2022 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// Using to test the concat interface of EcmaStringTable about intern string and non-intern string
let REGISTRY = Symbol();
function System() {
  this[REGISTRY] = {};
}
function getOrCreateLoad(loader, id) {
  let load = loader[REGISTRY][id];
  print("id: " + id + " - load: "+ load);
  load = loader[REGISTRY][id] = {};
  return load;
}

// non-intern + non-intern
let head1 = "no-".concat("schema:");
let tail1 = "/src/".concat("xxx-js/instantiated-1af0bf5b.js");
let key1 = head1 + tail1;
let key2 = head1 + tail1;

// intern + intern
let key3 = "no-schema:/src/xxx-js/instantiation.js";
let key4 = "no-schema:" + "/src/xxx-js/instantiation.js";

// non-intern + intern
let head2 = "no-".concat("schema:");
let tail2 = "/src/xxx-js/cc.js";
let key5 = "no-schema:" + "/src/xxx-js/cc.js";
let key6 = head2 + tail2;

// intern + non-intern
let head3 = "no-schema:";
let tail3 = "/src".concat("/instantiated-1af0bf5b.js");
let key7 = "no-schema:" + "/src/instantiated-1af0bf5b.js";
let key8 = head3 + tail3;

let keyArray = [key1, key3, key5, key7];
let system = new System();
for (let i = 0; i < keyArray.length; i++) {
    getOrCreateLoad(system, keyArray[i]);
}

print("key1 === key2: ", key1 === key2);
print("key3 === key4: ", key3 === key4);
print("key5 === key6: ", key5 === key6);
print("key7 === key8: ", key7 === key8);
getOrCreateLoad(system, key2);
getOrCreateLoad(system, key4);
getOrCreateLoad(system, key6);
getOrCreateLoad(system, key8);
